### PR TITLE
[Merged by Bors] - feat: improve typescript monorepo configuration (DX-37)

### PR DESCRIPTION
**Pull request by effervescentia**
**Fixes or implements DX-37**

### Brief description. What is this change?

Use the patterns developed for voiceflow/poc-ts-monorepo to improve the build time and type resolution of all packages in this monorepo.

### Implementation details. How do you make this change?

https://www.notion.so/voiceflow/TypeScript-Monorepo-POC-3c83192259334783a791fe7c5436042f

#### Special Notes

Since we need to support type checking in test files at development time (which was not covered in the POC setup) and only include the files from `src` at build time we have a `tsconfig.json`, `tsconfig.build.json` and potentially a `tsconfig.esm.json` in each package directory to cover each configuration

### Features

- roughly 25% faster build times locally and 40% faster on CI
- [dependency caching](https://www.notion.so/voiceflow/TypeScript-Monorepo-POC-3c83192259334783a791fe7c5436042f#970b089f70c64799a21d7d3bd6a6d38d)
- [type shortcutting](https://www.notion.so/voiceflow/TypeScript-Monorepo-POC-3c83192259334783a791fe7c5436042f#b2411f2599294310b0e8095555c5fc66)
- moved shared dev dependencies to the root `package.json`
- reduced complexity of `tsconfig*.json` files
- reduced complexity of `eslint` configuration
- use `eslint-output` for all sub-packages so that lint reporting on CI does not hide failures from the log

#### Before
---

build time locally:
Screen Shot 2022-07-13 at 11 52 09 PM

build time on ci:
Screen Shot 2022-07-14 at 12 41 16 AM

#### After
---

build time locally:
Screen Shot 2022-07-13 at 11 56 53 PM

when the cache hits 😎 locally:
Screen Shot 2022-07-14 at 4 20 26 AM

build time on ci:
Screen Shot 2022-07-14 at 12 39 23 AM

when the cache hits on ci:
Screen Shot 2022-07-14 at 4 15 56 AM

### Related

voiceflow/orb-common#73

### Checklist

- [ ] this is a breaking change and should publish a new major version
- [ ] appropriate tests have been written

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,32 +49,24 @@
...
-    "lint": "lerna run lint  --parallel",
-    "lint:fix": "lerna run lint:fix  --parallel",
-    "lint:quiet": "lerna run lint:quiet  --parallel",
-    "lint:report": "lerna run lint:report  --parallel",
-    "postinstall": "yarn bootstrap",
+    "lint": "lerna run lint --parallel",
+    "lint:fix": "lerna run lint:fix --parallel",
+    "lint:quiet": "lerna run lint:quiet --parallel",
+    "lint:report": "lerna run lint:report --parallel",
+    "postinstall": "[[ -e node_modules/_bootstrap ]] && lerna bootstrap && rm node_modules/_bootstrap || true",
```
Comment (effervescentia) [inline package.json:63]: now `lerna bootstrap` should only run on the initial install
if we add future packages we'll need to run `yarn bootstrap` to link them
however I've found having the `bootstrap` always run was occasionally causing issues with the install process

File: package.json [hunk 2]
```diff
@@ -35,32 +49,24 @@
...
-    "build:all": "yarn build",
-    "build:api-sdk": "yarn --cwd packages/api-sdk build",
-    "build:base-types": "yarn --cwd packages/base-types build",
-    "build:chat-types": "yarn --cwd packages/chat-types build",
-    "build:common": "yarn --cwd packages/common build",
-    "build:general-types": "yarn --cwd packages/general-types build",
-    "build:google-dfes-types": "yarn --cwd packages/google-dfes-types build",
-    "build:google-types": "yarn --cwd packages/google-types build",
-    "build:metrics": "yarn --cwd packages/metrics build",
-    "build:voice-types": "yarn --cwd packages/voice-types build",
```
Comment (effervescentia) [inline package.json:49]: this pattern wasn't being used or maintained

File: packages/nestjs-common/tsconfig.json [hunk 3]
```diff
@@ -1,33 +1,12 @@
...
-    "baseUrl": "./",
+    "baseUrl": ".",
+    "rootDir": "src",
+    "outDir": "build",
     "paths": {
       "@nestjs-common/*": ["src/*"]
-    },
-
-    // Strictness
-    "strict": true,
```
Comment (effervescentia) [inline packages/nestjs-common/tsconfig.json:15]: moved a lot of these up to our `@voiceflow/tsconfig/root.json` file which all other configs inherit from
removed flags we weren't using

**Review by jonahsnider — APPROVED** [review]

**Pull request closed without merging**